packages/yarnpkg-core/sources/scriptUtils.ts

```diff
@@ -272,7 +272,7 @@ export async function executePackageScript(locator: Locator, scriptName: string,
     const {manifest, env, cwd: realCwd} = await initializePackageEnvironment(locator, {project, binFolder, cwd, lifecycleScript: scriptName});
 
     const script = manifest.scripts.get(scriptName);
-    if (!script)
+    if (typeof script === `undefined`)
       return;
```

Comment:
I think it might rather be this:
```suggestion
    if (!script)
      return 0;
```

Reply:
Makes more sense not to execute an empty string, but I don't think we should return exit code 0 if the manifest doesn't have the specified script. 🤔

Reply:
Thinking more, I think you're right. I have a monitoring plugin tracking the scripts that execute on my project, and I'd probably consider it a bug if empty ones weren't reported. Good thinking 👍

That being said, there's still a bug somewhere: the `exitCode` variable in `Project.ts` should be typed as `number`, which would have allowed us to catch this problem earlier. Can you make the fix in the same PR? I wonder if there's a lint rule that could prevent variable declarations that are missing both explicit and inferred types 🤔

Reply:
There's [`explicit-function-return-type`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/explicit-function-return-type.md) in `@typescript-eslint/eslint-plugin` that can prevent this kind of errors, but it requires always annotating the return type. I'd personally enable it, it's even part of their recommended config. Most importantly, it makes us think twice before annotating the return type. What do you think?

Reply:
I meant [this code](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-core/sources/Project.ts#L1319-L1328). Since the `exitCode` declaration doesn't have explicit type, nor initializer, TS silently types it as `any`, which allowed `undefined` to be returned. Maybe [`no-unsafe-assignment`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unsafe-assignment.md)?

I don't have strong opinions on explicit function return types, except it would probably require a bunch of tedious changes 🤔 Maybe it'd be simpler to just ban functions from mixing returns with and without values, which is easy to do with [`consistent-return`](consistent-return).